File: src/Samples/AvaloniaXPlatExample/AvaloniaXPlatExample/ViewModels/CounterViewModel.fs
namespace AvaloniaXPlatExample.ViewModels

open System
open Elmish.Avalonia
open Elmish

module Counter = 
    type Model =  { Count: int; Actions: Action list }
    and Action = { Description: string; Timestamp: DateTime }

    type Msg = 
        | Increment
        | Decrement
        | Reset

    let init() = 
        { 
            Count = 0
            Actions = [ { Description = "Initialized count."; Timestamp = DateTime.Now } ]
        }

    let update (msg: Msg) (model: Model) = 
        match msg with
        | Increment ->
            { 
                Count = model.Count + 1 
                Actions = model.Actions @ [ { Description = "Incremented"; Timestamp = DateTime.Now } ]
            }
        | Decrement ->
            { 
                Count = model.Count - 1 
                Actions = model.Actions @ [ { Description = "Decremented"; Timestamp = DateTime.Now } ] 
            }
        | Reset ->
            init()


open Counter

type CounterViewModel() =
    inherit ReactiveElmishViewModel()

    let local = 
        Program.mkAvaloniaSimple init update
        //|> Program.withConsoleTrace
        |> Program.mkStore

    member this.Count = this.Bind(local, _.Count)
    member this.Actions = this.Bind(local, _.Actions)
    member this.Increment() = local.Dispatch Increment
    member this.Decrement() = local.Dispatch Decrement
    member this.Reset() = local.Dispatch Reset
    member this.IsResetEnabled = this.Bind(local, fun m -> m.Count <> 0)

    static member DesignVM = 
        new CounterViewModel() 

File: src/WpfExample/ViewModels/CounterViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CommunityToolkit.Mvvm.Input;
using ReactiveElmish;


using ReactiveUI;

namespace WpfExample.ViewModels;

public class CounterStore() : ReactiveStore<CounterModel>(CounterModel.Init())
{
    public void Increment() => 
        Update(m => m with { Count = m.Count + 1, Actions = m.Actions.Append(new Action("Increment", DateTime.Now)).ToArray() });

    public void Decrement() => 
        Update(m => m with { Count = m.Count - 1, Actions = m.Actions.Append(new Action("Decrement", DateTime.Now)).ToArray() });

    public void Reset() => 
        Update(m => CounterModel.Init());
}

public record CounterModel(int Count, Action[] Actions) 
{
    public static CounterModel Init() => new CounterModel(0, [new Action("Initialized", DateTime.Now)]);
}

public record Action(string Description, DateTime Timestamp);

public class CounterViewModel : ReactiveObject
{
    readonly CounterStore store = new CounterStore();
    public CounterViewModel() => Rx = new ReactiveBindingsCS(this.RaisePropertyChanged);                
    ReactiveBindingsCS Rx { get; }

    public int Count => Rx.Bind(store, s => s.Count);
    public ReadOnlyCollection<Action> Actions => Rx.BindList(store, s => s.Actions);
    public IRelayCommand Increment { get => new RelayCommand(store.Increment); }
    public IRelayCommand Decrement { get => new RelayCommand(store.Decrement); }
    public IRelayCommand Reset { get => new RelayCommand(store.Reset); }
}
